text/0000-local-template-blocks.md

```diff
@@ -0,0 +1,133 @@
+- Ember Issue: (leave this empty)
+
+# Summary
+
+There is strong demand for the ability to pass in template blocks into higher components as a means to configure/override how/what they render to DOM, but unfortunately other proposals to that effect have failed to gain traction. This RFC proposes a more general-purpose syntax for defining locally-named template blocks that can be passed into higher order components as attrs.
+
+# Motivation
+
+The need/demand for "named yields"/"block slots" is well-established:
+
+- https://github.com/emberjs/rfcs/pull/43
+- https://github.com/emberjs/rfcs/pull/72
+- https://github.com/emberjs/rfcs/pull/193
+- https://github.com/ciena-blueplanet/ember-block-slots
+- https://github.com/ryanto/ember-content-for
+
+In short, people want a way to pass in more than just the default/inverse template blocks, an unfortunate limitation to the flexibility/power of higher order components, and which generally speaking hurts the composability story of Ember components.
+
+The problem with all of the above approaches (aside from the fact that they hack around the lack of such an officially-supported syntax/approach) is that they miss an important and more general use case, which is the ability to define a named local template block and pass it to multiple components, e.g.:
+
+```hbs
+{{#let-block sharedHeader as |person|}}
+  {{person.name}}
+{{/let-block}}
+
+{{complex-component
+  header=sharedHeader}}
+{{complex-component
+  header=sharedHeader}}
+{{complex-component
+  header=sharedHeader}}
+```
+
+None of the proposals I'm aware of support defining `sharedHeader` and passing it to multiple separate components; the most they can do is pass a named block to a higher order component and let the higher order component render it multiple times. It's a subtle but important limitation.
+
+So, given this limitation, and that these proposals have languished, I suggest we ship a more basic / verbose but more fully-featured / flexible syntax today and build a nicer block-slot syntax based off of it in a separate RFC.
+
+# Detailed design
+
+I propose two new additions / primitives to the present-day templating API:
+
+## 1. let-block
+
+`let-block` is a Glimmer built-in that defines a local variable/constant within the current lexical scope that holds a reference to a block that can be rendered multiple times.
+
+```hbs
+{{#let-block fooBlock as |a b c|}}
+  Hello {{a}}, {{b}}, and {{c}}!
+{{/let-block}}
+
+{{x-bar headerBlock=fooBlock}}
+{{x-wat headerBlock=fooBlock footerBlock=fooBlock}}
+```
+
+`let-block` declarations behave similar to block params in that a reference to `fooBlock` in the above template will always reference the block declared by `let-block` and never try and do a property lookup of `fooBlock` on the context.
+
+`let-block` declarations are hoisted to the top of the current lexical scope, like classic JS function declarations; this means it's possible to reference a block _above_ where it is defined.
```

Comment:
What's a use case for hoisting `let-block`s?

> this means it's possible to reference a block _above_ where it is defined

This seems like it might be confusing as it is inconsistent with JavaScript `let` hoisting.